src/Optimization/hiopKKTLinSysSparseCondensed.cpp

```diff
@@ -0,0 +1,1245 @@
+  KKTmat->makeCompressed();
+  t.stop();
+  if(perf_report_) nlp_->log->printf(hovSummary, "makeCompressed         %.3f sec\n", t.getElapsedTime());
+
+  return KKTmat;
+} 
+#endif //HIOP_USE_EIGEN
+/*******************************************************************************************************
+ * hiopMatrixSparseCSRStorage
+ *******************************************************************************************************/
+hiopMatrixSparseCSRStorage::hiopMatrixSparseCSRStorage()
+  : nrows_(0),
+    ncols_(0),
+    nnz_(0),
+    irowptr_(nullptr),
+    jcolind_(nullptr),
+    values_(nullptr)
+{
+  
+}
+hiopMatrixSparseCSRStorage::hiopMatrixSparseCSRStorage(size_type m, size_type n, size_type nnz)
+  : nrows_(m),
+    ncols_(n),
+    nnz_(nnz),
+    irowptr_(nullptr),
+    jcolind_(nullptr),
+    values_(nullptr)
+{
+  alloc();
+}
+hiopMatrixSparseCSRStorage::~hiopMatrixSparseCSRStorage()
+{
+  dealloc();
+}
+
+void hiopMatrixSparseCSRStorage::alloc()
+{
+  assert(irowptr_ == nullptr);
+  assert(jcolind_ == nullptr);
+  assert(values_ == nullptr);
+
+  irowptr_ = new index_type[nrows_+1];
+  jcolind_ = new index_type[nnz_];
+  values_ = new double[nnz_];
+}
+
+
+void hiopMatrixSparseCSRStorage::dealloc()
+{
+  delete[] irowptr_;
+  delete[] jcolind_;
+  delete[] values_;
+  irowptr_ = nullptr;
+  jcolind_ = nullptr;
+  values_ = nullptr;
+}
+
+/**
+ * Forms a CSR matrix from a sparse matrix in triplet format. Returns false if the input formated 
+ * as expected (e.g., ordered by rows then by columns), otherwise returns true.
+*/
+bool hiopMatrixSparseCSRStorage::form_from(const hiopMatrixSparseTriplet& M)
+{
+  if(M.m()!=nrows_ || M.n()!=ncols_ || M.numberOfNonzeros()!=nnz_) {
+    dealloc();
+    
+    nrows_ = M.m();
+    ncols_ = M.n();
+    nnz_ = M.numberOfNonzeros();
+
+    alloc();
+  }
+
+  assert(nnz_>=0);
+  if(nnz_<=0) {
+    return true;
+  }
+  
+  assert(irowptr_);
+  assert(jcolind_);
+  assert(values_);
+
+  const index_type* Mirow = M.i_row();
+  const index_type* Mjcol = M.j_col();
+  const double* Mvalues  = M.M();
+
+  //storage the row count
+  index_type w[nrows_];
+  for(int i=0; i<nrows_; ++i) {
+    w[i] = 0;
+  }
+  
+  for(int it=0; it<nnz_; ++it) {
+    const index_type row_idx = Mirow[it];
+
+#ifndef NDEBUG
+    if(it>0) {
+      assert(Mirow[it] >= Mirow[it-1] && "row indexes of the triplet format are not ordered.");
+      if(Mirow[it] == Mirow[it-1]) {
+        assert(Mjcol[it] > Mjcol[it-1] && "col indexes of the triplet format are not ordered or unique.");
+      }
+    }
+#endif
+    assert(row_idx<nrows_ && row_idx>=0);
+    assert(Mjcol[it]<ncols_ && Mjcol[it]>=0);
+
+    w[row_idx]++;
+
+    jcolind_[it] = Mjcol[it];
+    values_[it] = Mvalues[it];
+  }
+
+  irowptr_[0] = 0;
+  for(int i=0; i<nrows_; i++) {
+    irowptr_[i+1] = irowptr_[i] + w[i];
+  }
+  assert(irowptr_[nrows_] == nnz_);
+  return true;
+}
+
+bool hiopMatrixSparseCSRStorage::form_from(const size_type m,
```

Comment:
Seems like this is almost identical to the other `form_from`. Could we just have the `form_from` that takes a matrix just pull the relevant values from the parameter and then call this member function?

Reply:
the two methods are a bit different as only one updates the values. The code is temporary and will be refactored when we move to a matrix class.